src/ast/passes/semantic_analyser.cpp

```diff
@@ -37,9 +37,53 @@ static const std::map<std::string, std::tuple<size_t, bool>> &getIntcasts()
   return intcasts;
 }
 
+template <typename T>
+static std::pair<T, T> getIntTypeRange(const SizedType &ty)
+{
+  assert(ty.IsIntegerTy());
+  auto size = ty.GetSize();
+  if (size == 1) {
+    if (ty.IsSigned()) {
+      return { CHAR_MIN, CHAR_MAX };
+    } else {
+      return { 0, UCHAR_MAX };
+    }
+  } else if (size == 2) {
+    if (ty.IsSigned()) {
+      return { SHRT_MIN, SHRT_MAX };
+    } else {
+      return { 0, USHRT_MAX };
+    }
+  } else if (size == 4) {
+    if (ty.IsSigned()) {
+      return { INT_MIN, INT_MAX };
+    } else {
+      return { 0, UINT_MAX };
+    }
+  } else if (size == 8) {
+    if (ty.IsSigned()) {
+      return { LONG_MIN, LONG_MAX };
+    } else {
+      return { 0, ULONG_MAX };
```

Comment:
We'd also be better off using `std::numeric_limits<T>::max` in place of the platform-dependent C macros here